### Cristeam Caiola 


### Dataset Description

- excel file provided in online course


In [177]:
from keras import optimizers
import numpy as np
from keras.models import Sequential
from keras.layers import Activation, Dense
from keras.layers import Dropout
from sklearn.model_selection import train_test_split
from keras.layers import BatchNormalization

In [ ]:

#Load data
X_train = np.loadtxt('w4_train_data.csv', delimiter=',', skiprows=1)
y_train= np.loadtxt('w4_train_data - y.csv', delimiter=',', skiprows=1) 
X_test= np.loadtxt("w4_test_data.csv", delimiter=',', skiprows=1)
y_test = np.loadtxt("w4_test_data - y.csv", delimiter=',', skiprows=1)
print("------------------------------------------------------- ")
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)


### 1. Creating a model
- Keras model object can be created with Sequential class
- At the outset, the model is empty per se. It is completed by **'adding'** additional layers and compilation
- Doc: https://keras.io/models/sequential/

In [178]:
model = Sequential()

### 1-1. Adding layers
- Keras layers can be **added** to the model
- Adding layers are like stacking lego blocks one by one
- Doc: https://keras.io/layers/core/

In [179]:
# Keras model with two hidden layer with 10 neurons each 
model.add(Dense(10, input_shape = (784,)))    # Input layer => input_shape should be explicitly designated
model.add(BatchNormalization())                    # Add Batchnorm layer before Activation
model.add(Activation('sigmoid'))
#model.add(Dropout(0.2))                        # Dropout layer after Activation
model.add(Dense(10))                         # Hidden layer => only output dimension should be designated
model.add(BatchNormalization())                    # Add Batchnorm layer before Activation
model.add(Activation('sigmoid'))
#model.add(Dropout(0.2))                        # Dropout layer after Activation
model.add(Dense(10))                         # Hidden layer => only output dimension should be designated
model.add(BatchNormalization())                    # Add Batchnorm layer before Activation
model.add(Activation('sigmoid'))
#model.add(Dropout(0.2))                        # Dropout layer after Activation
model.add(Dense(1))                          # Output layer => output dimension = 1 since it is regression problem


In [180]:
# This is equivalent to the above code block
#model.add(Dense(10, input_shape = (784,), activation = 'sigmoid'))
#model.add(Dense(10, activation = 'sigmoid'))
#model.add(Dense(10, activation = 'sigmoid'))
#model.add(Dense(1))

### 1-2. Model compile
- Keras model should be "compiled" prior to training
- Types of loss (function) and optimizer should be designated
    - Doc (optimizers): https://keras.io/optimizers/
    - Doc (losses): https://keras.io/losses/

In [181]:
#Optimazer
sgd = optimizers.SGD(lr = 0.001)    # stochastic gradient descent optimizer
#adam = optimizers.Adam(lr = 0.001)    # use Adam optimizer

In [182]:
#compile of model
model.compile(optimizer = sgd, loss = 'mean_squared_error', metrics = ['mse'])    # for regression problems, mean squared error (MSE) is often employed
#model.compile(optimizer = adam, loss = 'categorical_crossentropy', metrics = ['accuracy'])


### Summary of the model

In [183]:
model.summary()

Model: "sequential_9"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_47 (Dense)             (None, 10)                7850      
_________________________________________________________________
batch_normalization_1 (Batch (None, 10)                40        
_________________________________________________________________
activation_30 (Activation)   (None, 10)                0         
_________________________________________________________________
dense_48 (Dense)             (None, 10)                110       
_________________________________________________________________
batch_normalization_2 (Batch (None, 10)                40        
_________________________________________________________________
activation_31 (Activation)   (None, 10)                0         
_________________________________________________________________
dense_49 (Dense)             (None, 10)               

### 2. Training
- Training the model with training data provided

In [184]:
model.fit(X_train, y_train, batch_size = 50, epochs = 100, verbose = 1)

Epoch 1/100
20580/20580 [==============================] - 1s 53us/step - loss: 10.4405 - mse: 10.4405
Epoch 2/100
20580/20580 [==============================] - 1s 34us/step - loss: 4.3572 - mse: 4.3572
Epoch 3/100
20580/20580 [==============================] - 1s 35us/step - loss: 3.6751 - mse: 3.6751
Epoch 4/100
20580/20580 [==============================] - 1s 37us/step - loss: 3.3572 - mse: 3.3572
Epoch 5/100
20580/20580 [==============================] - 1s 38us/step - loss: 3.1486 - mse: 3.1486
Epoch 6/100
20580/20580 [==============================] - 1s 36us/step - loss: 2.9952 - mse: 2.9952
Epoch 7/100
20580/20580 [==============================] - 1s 35us/step - loss: 2.8162 - mse: 2.8162
Epoch 8/100
20580/20580 [==============================] - 1s 33us/step - loss: 2.7329 - mse: 2.7329
Epoch 9/100
20580/20580 [==============================] - 1s 36us/step - loss: 2.6316 - mse: 2.6316
Epoch 10/100
20580/20580 [==============================] - 1s 36us/step - loss: 2.5470 -

20580/20580 [==============================] - 1s 34us/step - loss: 1.0876 - mse: 1.0876
Epoch 82/100
20580/20580 [==============================] - 1s 36us/step - loss: 1.1032 - mse: 1.1032
Epoch 83/100
20580/20580 [==============================] - 1s 37us/step - loss: 1.0702 - mse: 1.0702
Epoch 84/100
20580/20580 [==============================] - 1s 36us/step - loss: 1.0570 - mse: 1.0570
Epoch 85/100
20580/20580 [==============================] - 1s 35us/step - loss: 1.0659 - mse: 1.0659
Epoch 86/100
20580/20580 [==============================] - 1s 40us/step - loss: 1.0671 - mse: 1.0671
Epoch 87/100
20580/20580 [==============================] - 1s 50us/step - loss: 1.0508 - mse: 1.0508
Epoch 88/100
20580/20580 [==============================] - 1s 50us/step - loss: 1.0276 - mse: 1.0276
Epoch 89/100
20580/20580 [==============================] - 1s 49us/step - loss: 1.0151 - mse: 1.0151
Epoch 90/100
20580/20580 [==============================] - 1s 42us/step - loss: 1.0319 - mse: 

### 3. Evaluation
- Keras model can be evaluated with evaluate() function
- Evaluation results are contained in a list
    - Doc (metrics): https://keras.io/metrics/

In [185]:
results = model.evaluate(X_test, y_test)

8820/8820 [==============================] - 0s 37us/step


In [186]:
print(model.metrics_names)     # list of metric names the model is employing
print(results)                 # actual figure of metrics computed


['loss', 'mse']
[1.0298939371595577, 1.0298937559127808]


In [187]:
print('loss: ', results[0])
print('mse: ', results[1])

loss:  1.0298939371595577
mse:  1.0298937559127808


# Results 

<b>1-without optimization and SGD=0.01: <br></b>
<u>-Training:</u> <br>
Epoch 100/100 60000/60000 [==============================] - 1s 22us/step - loss: 3.5553 - mse: 3.5553
        
<u>-Evaluation:</u><br>
loss:  3.2445113330841067
mse:  3.244511604309082

<b>2-with Dropout and SGD = 0.01</b><br>
<u>-Training: </u><br>
Epoch 100/100
 60000/60000 [==============================] - 2s 30us/step - loss: 8.3513 - mse: 8.3513

<u>-Evaluation:</u><br>
loss:  8.386413145446777<br>
mse:  8.386411666870117

<b>3-with Dropout and SGD = 0.001</b><br>
<u>-Training: </u><br>
Epoch 100/100 29400/29400 [==============================] - 1s 29us/step - loss: 3.0107 - mse: 3.0107

<u>-Evaluation:</u><br>
loss:  2.0790235421771093<br>
mse:  2.0790231227874756

<b>4-With Batch Normalization</b><br>
<u>-Training: </u><br>
Epoch 100/100 20580/20580 [==============================] - 1s 33us/step - loss: 0.9865 - mse: 0.9864

<u>-Evaluation:</u><br>
loss:  1.0298939371595577 <br>
mse:  1.0298937559127808


## MLP for classification tasks
- When the target (**y**) is discrete (categorical)
- For loss function, cross-entropy is used and for evaluation metric, accuracy is commonly used

In [188]:
X_train, X_test, y_train, y_test = train_test_split(X_train, y_train, test_size = 0.3, random_state = 7) 

### Dataset Description
- Breast cancer dataset has total 569 data instances (212 malign, 357 benign instances)
- 30 attributes (features) to predict the binary class (M/B)
- Doc: http://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_breast_cancer.html#sklearn.datasets.load_breast_cancer

In [189]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)


(14406, 784)
(6174, 784)
(14406,)
(6174,)


### 1. Creating a model
- Same with regression model at the outset

In [190]:
model = Sequential()

### 1-1. Adding layers
- Keras layers can be **added** to the model
- Adding layers are like stacking lego blocks one by one
- It should be noted that as this is a classification problem, sigmoid layer (softmax for multi-class problems) should be added
- Doc: https://keras.io/layers/core/

In [191]:
# Keras model with two hidden layer with 10 neurons each 
model.add(Dense(10, input_shape = (784,)))    # Input layer => input_shape should be explicitly designated
model.add(BatchNormalization())                    # Add Batchnorm layer before Activation
model.add(Activation('sigmoid'))
#odel.add(Dropout(0.2))
model.add(Dense(10))                         # Hidden layer => only output dimension should be designated
model.add(BatchNormalization())                    # Add Batchnorm layer before Activation
model.add(Activation('sigmoid'))
#odel.add(Dropout(0.2))                        # Dropout layer after Activation
model.add(Dense(10))                         # Hidden layer => only output dimension should be designated
model.add(BatchNormalization())                    # Add Batchnorm layer before Activation
model.add(Activation('sigmoid'))
#odel.add(Dropout(0.2))                        # Dropout layer after Activation
model.add(Dense(1))                          # Output layer => output dimension = 1 since it is regression problem
model.add(Activation('sigmoid'))

In [192]:
# This is equivalent to the above code block
#model.add(Dense(10, input_shape = (784,), activation = 'sigmoid'))
#model.add(Dense(10, activation = 'sigmoid'))
#model.add(Dense(10, activation = 'sigmoid'))
#model.add(Dense(1, activation = 'sigmoid'))

### 1-2. Model compile
- Keras model should be "compiled" prior to training
- Types of loss (function) and optimizer should be designated
    - Doc (optimizers): https://keras.io/optimizers/
    - Doc (losses): https://keras.io/losses/

In [193]:
#from keras.utils import to_categorical

In [194]:
sgd = optimizers.SGD(lr = 0.001)    # stochastic gradient descent optimizer
#adam = optimizers.Adam(lr = 0.001)    # use Adam optimizer

In [195]:
model.compile(optimizer = sgd, loss = 'binary_crossentropy', metrics = ['accuracy'])
#model.compile(optimizer = adam, loss = 'categorical_crossentropy', metrics = ['accuracy'])


### Summary of the model

In [196]:
model.summary()

Model: "sequential_10"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_51 (Dense)             (None, 10)                7850      
_________________________________________________________________
batch_normalization_4 (Batch (None, 10)                40        
_________________________________________________________________
activation_33 (Activation)   (None, 10)                0         
_________________________________________________________________
dense_52 (Dense)             (None, 10)                110       
_________________________________________________________________
batch_normalization_5 (Batch (None, 10)                40        
_________________________________________________________________
activation_34 (Activation)   (None, 10)                0         
_________________________________________________________________
dense_53 (Dense)             (None, 10)              

### 2. Training
- Training the model with training data provided

In [197]:
model.fit(X_train, y_train, batch_size = 50, epochs = 100, verbose = 1)

Epoch 1/100
14406/14406 [==============================] - 1s 63us/step - loss: -7.2544 - accuracy: 0.1127
Epoch 2/100
14406/14406 [==============================] - 1s 35us/step - loss: -23.5853 - accuracy: 0.1092
Epoch 3/100
14406/14406 [==============================] - 1s 36us/step - loss: -43.9062 - accuracy: 0.1092
Epoch 4/100
14406/14406 [==============================] - 1s 36us/step - loss: -69.7902 - accuracy: 0.1092
Epoch 5/100
14406/14406 [==============================] - 1s 42us/step - loss: -100.2065 - accuracy: 0.1092
Epoch 6/100
14406/14406 [==============================] - 1s 37us/step - loss: -133.8898 - accuracy: 0.1092
Epoch 7/100
14406/14406 [==============================] - 1s 40us/step - loss: -170.1446 - accuracy: 0.1092
Epoch 8/100
14406/14406 [==============================] - 1s 38us/step - loss: -207.9028 - accuracy: 0.1092
Epoch 9/100
14406/14406 [==============================] - 1s 39us/step - loss: -246.3727 - accuracy: 0.1092
Epoch 10/100
14406/14406

14406/14406 [==============================] - 1s 36us/step - loss: -2868.8289 - accuracy: 0.1092
Epoch 76/100
14406/14406 [==============================] - 1s 35us/step - loss: -2908.3583 - accuracy: 0.1092
Epoch 77/100
14406/14406 [==============================] - 1s 39us/step - loss: -2948.0674 - accuracy: 0.1092
Epoch 78/100
14406/14406 [==============================] - 1s 35us/step - loss: -2987.8762 - accuracy: 0.1092
Epoch 79/100
14406/14406 [==============================] - 1s 35us/step - loss: -3027.3646 - accuracy: 0.1092
Epoch 80/100
14406/14406 [==============================] - 0s 34us/step - loss: -3063.6032 - accuracy: 0.1092
Epoch 81/100
14406/14406 [==============================] - 0s 35us/step - loss: -3090.5192 - accuracy: 0.1092
Epoch 82/100
14406/14406 [==============================] - 1s 36us/step - loss: -3145.5962 - accuracy: 0.1092
Epoch 83/100
14406/14406 [==============================] - 0s 35us/step - loss: -3186.4306 - accuracy: 0.1092
Epoch 84/100
1

### 3. Evaluation
- Keras model can be evaluated with evaluate() function
- Evaluation results are contained in a list
    - Doc (metrics): https://keras.io/metrics/

In [198]:
results = model.evaluate(X_test, y_test)

6174/6174 [==============================] - 0s 46us/step


In [199]:
print(model.metrics_names)     # list of metric names the model is employing
print(results)                 # actual figure of metrics computed

['loss', 'accuracy']
[-3879.7241630887847, 0.11305474489927292]


In [200]:
print('loss: ', results[0])
print('accuracy: ', results[1])

loss:  -3879.7241630887847
accuracy:  0.11305474489927292


# Results

<b>1-without optimization: </b><br> 
<u>Training:</u> <br>
Epoch 100/100 
42000/42000 [==============================] - 1s 28us/step - loss: -109484.1585 - accuracy: 0.1118

<u>Evaluation:</u><br>
loss:  -110271.03106944445
accuracy:  0.11366666853427887

<b>2-with Dropout and SGD = 0,01</b><br>
<u>Training:</u> <br>
Epoch 100/100
29400/29400 [==============================] - 1s 31us/step - loss: -77036.9652 - accuracy: 0.1102
    
<u>Evaluation:</u><br>
loss:  -76669.99157242064<br>
accuracy:  0.11547619104385376


<b>3-with Dropout and SGD = 0,001</b><br>
<u>Training:</u> <br>
Epoch 100/100 20580/20580 [==============================] - 1s 33us/step - loss: -5331.8802 - accuracy: 0.1103

<u>Evaluation:</u><br>
loss:  -5388.151646648243<br>
accuracy:  0.10997732728719711


<b>4-With Batch Normalization</b><br>
<u>-Training: </u><br>
Epoch 100/100 14406/14406 [==============================] - 1s 36us/step - loss: -3865.8678 - accuracy: 0.1092

<u>-Evaluation:</u><br>
loss:  -3879.7241630887847<br>
accuracy:  0.11305474489927292